### PR TITLE
👀 [Frontend] 282 adding initial instrumentation setup

### PR DESCRIPTION
**Pull request by xoscar**
Fixes https://github.com/open-telemetry/opentelemetry-demo/issues/283.

## Changes

Adds the instrumentation code for the front-end application. Changes include:
1. Adding auto instrumentation and propagation for the browser side.
2. Adding manual and auto instrumentation for the server (node.js) side.

https://www.loom.com/share/81c1c009c7e3492087ceecf8a0c6f21d

## Note
This is just the initial instrumentation wire up, for quality attributes, spans, etc we can have a discussion later on

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -7,6 +7,8 @@ OTEL_EXPORTER_OTLP_ENDPOINT=http://otelcol:4317
 OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=${OTEL_EXPORTER_OTLP_ENDPOINT}
 OTEL_EXPORTER_OTLP_METRICS_ENDPOINT=${OTEL_EXPORTER_OTLP_ENDPOINT}
 
+OTEL_EXPORTER_OTLP_TRACES_PUBLIC_ENDPOINT=http://localhost:4318/v1/traces
```
Comment (xoscar) [inline .env:10]: Adding thew OTEL public endpoint for browser connection

File: docker-compose.yml [hunk 2]
```diff
@@ -30,7 +30,7 @@ services:
       - ./src/otelcollector/otelcol-config-extras.yml:/etc/otelcol-config-extras.yml
     ports:
       - "4317"          # OTLP over gRPC receiver
-      - "4318"          # OTLP over HTTP receiver
+      - "4318:4318"          # OTLP over HTTP receiver
```
Comment (xoscar) [inline docker-compose.yml:33]: Exposing the http port from the otel collector

File: src/frontend/Dockerfile [hunk 3]
```diff
@@ -37,4 +39,4 @@ USER nextjs
 ENV PORT 8080
 EXPOSE ${PORT}
 
-CMD ["node", "server.js"]
+CMD ["sh", "-c", "npm start"]
```
Comment (xoscar) [inline src/frontend/Dockerfile:42]: Updating this to use a package.json script

File: src/frontend/next.config.js [hunk 4]
```diff
@@ -45,7 +49,10 @@ const nextConfig = {
     PRODUCT_CATALOG_SERVICE_ADDR,
     RECOMMENDATION_SERVICE_ADDR,
     SHIPPING_SERVICE_ADDR,
+    OTEL_EXPORTER_OTLP_TRACES_ENDPOINT,
```
Comment (xoscar) [inline src/frontend/next.config.js:52]: Passing the env variables to the next context

File: src/frontend/package.json [hunk 5]
```diff
@@ -3,38 +3,49 @@
   "version": "0.1.0",
   "private": true,
   "scripts": {
-    "dev": "npm run grpc:generate && next dev",
+    "dev": "npm run grpc:generate && NODE_OPTIONS='-r ./utils/telemetry/Instrumentation.js' next dev",
```
Comment (xoscar) [inline src/frontend/package.json:6]: requiring the Instrumentation script before running

File: src/frontend/pages/api/cart.ts [hunk 6]
```diff
@@ -49,4 +50,4 @@ const handler = async ({ method, body, query }: NextApiRequest, res: NextApiResp
   }
 };
 
-export default handler;
+export default InstrumentationMiddleware(handler);
```
Comment (xoscar) [inline src/frontend/pages/api/cart.ts:53]: Every api route is now wrapped with this High Order Functions which provides the manual instrumentation

File: src/frontend/utils/telemetry/BackendTracer.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { context, trace, Tracer, Span, Context, SpanOptions, SpanStatusCode, Exception } from '@opentelemetry/api';
+
+interface ITracer {
```
Comment (xoscar) [inline src/frontend/utils/telemetry/BackendTracer.ts:3]: Provides utility functions for the instrumentation

File: src/frontend/utils/telemetry/FrontendTracer.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import { CompositePropagator, W3CBaggagePropagator, W3CTraceContextPropagator } from '@opentelemetry/core';
```
Comment (xoscar) [inline src/frontend/utils/telemetry/FrontendTracer.ts:1]: Creates the provider and context zones for the frontend implementation

File: src/frontend/utils/telemetry/InstrumentationMiddleware.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import { NextApiHandler } from 'next';
+import Tracer from './BackendTracer';
+import { context, propagation, SpanKind, SpanStatusCode, Exception } from '@opentelemetry/api';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+
+const InstrumentationMiddleware = (handler: NextApiHandler): NextApiHandler => {
```
Comment (xoscar) [inline src/frontend/utils/telemetry/InstrumentationMiddleware.ts:6]: This is the HOF that generates the manual span for each incoming request

File: src/otelcollector/otelcol-config.yml [hunk 10]
```diff
@@ -3,6 +3,9 @@ receivers:
     protocols:
       grpc:
       http:
+        cors:
+          allowed_origins:
```
Comment (xoscar) [inline src/otelcollector/otelcol-config.yml:7]: Allowing CORS for the OTEL collection for the HTTP browser side requests

**Review by puckpuck — CHANGES_REQUESTED** [review]

**Comment by xoscar** [comment]
Hey @puckpuck thanks for the review! 😄 I just updated the PR with your comments

**Comment by xoscar** [comment]
Hey @puckpuck I just updated the PR with a fix for the error you were seeing, we still need to fix the currency service issue but I think that can be done in a separate PR as the problem exists in the main branch.

The problem was that we instantiated two zone managers and Next.js already has one.

**Comment by puckpuck** [comment]
@xoscar, now that we fixed the other services, can you update this branch so we have a clean setup to work from?

**Comment by xoscar** [comment]
@puckpuck  done 😃

**Review by mic-max — APPROVED** [review]

**Comment by puckpuck** [comment]
Root spans for traces initiated from the browser are missing.
Screen Shot 2022-08-15 at 11 38 52 PM

Using chrome dev tools, you can see 405 errors to send a payload of what seems to be an OTLP JSON payload.

Screen Shot 2022-08-15 at 11 54 10 PM
Screen Shot 2022-08-15 at 11 54 39 PM

**Comment by xoscar** [comment]
@puckpuck the latest commit I submitted has some fixes for that problem, it should now have the correct root span starting from the client side 😄

**Review by puckpuck — APPROVED** [review]
Tested it end to end on my side. Works wonderfully!

**Comment by cartersocha** [comment]
@xoscar update this branch & we'll merge next

**Comment by xoscar** [comment]
@cartersocha Done 😉

**Comment by cartersocha** [comment]
Did you implement Michael's suggestions? I'm about to resolve the conversations. 

Also please add a change log entry 😀

**Comment by xoscar** [comment]
@cartersocha Yes and Done

**Comment by cartersocha** [comment]
FYI @austinlparker you might want to pull latest for your demo

**Pull request merged**